storage: add new backend `HttpProxy`

loheagn opened the pull request:
This pr adds a new storage backend `HttpProxy` which can access blobs through a http proxy server.

The http proxy server can be local (using unix socket) or remote (using `https://` or using `http://`).

`HttpProxy` uses two API endpoints to access the blobs:
- `HEAD /path/to/blobs` to get the blob size
- `GET /path/to/blobs` to read the blob

The http proxy server should respect [the `Range` header](https://www.rfc-editor.org/rfc/rfc9110.html#name-range) to compute the offset and length of the blob.

The example config files for this new backend may be:

```jsonc
// for remote usage
{
  "backend": {
    "type": "http-proxy",
    "config": {
      "addr": "http://127.0.0.1:9977",
      "path": "/namespace/<repo>/blobs"
    }
  }
}
```

or

```jsonc
// for local unix socket
{
  "backend": {
    "type": "http-proxy",
    "config": {
      "addr": "/path/to/unix/socket/file"
    }
  }
}
```

There is also a test in `http_proxy.rs` to make sure `HttpProxy` works well, which setups a simple http server and generates a `HttpProxy` backend to get contents from the server.

Signed-off-by: Nan Li <loheagn@icloud.com>

imeoer commented:
Thanks @loheagn for the contribution, it can make `nydus-image unpack` faster!

My suggestion is that maybe we should implement the content store backend with HTTP over unix domain socket instead of TCP, which is simpler and requires only implementing two distribution APIs:

- `HEAD` blob size;
- `GET` blob data (HTTP range get);

It's very like `contrib/nydus-backend-proxy` server (it receives HTTP requests and proxy to a local blob directory).

jiangliu commented on api/src/config.rs:
```diff
@@ -427,6 +441,14 @@ pub struct S3Config {
 
+/// ContentProxyProxy configuration information to access blobs.
+#[derive(Clone, Debug, Default, Deserialize, Eq, PartialEq, Serialize)]
+pub struct LocalHttpProxyConfig {
```
How about `LocalUdsConfig`?
We can define an APIs to retrieve blobs from Unix domain socket, no matter it's containerd content store or something else.

jiangliu replied:
Or `ContainerdContentStoreProxy`? @imeoer

loheagn replied:
But this backend driver only support http protocol now. I think `LocalHttpProxy` may be better.

imeoer replied:
It's ok for me if using `LocalHttpProxy`, maybe we can merge the feature with [nydus-backend-proxy](https://github.com/dragonflyoss/image-service/tree/master/contrib/nydus-backend-proxy) (not in this PR)? So that we can use `LocalHttpProxy` to request an HTTP server that can proxy to local blob directory/containerd content store.

imeoer commented on api/src/config.rs:
```diff
@@ -427,6 +441,14 @@ pub struct S3Config {
+/// ContentProxyProxy configuration information to access blobs.
+#[derive(Clone, Debug, Default, Deserialize, Eq, PartialEq, Serialize)]
+pub struct LocalHttpProxyConfig {
+    pub socket_path: String,
```
Maybe we can name it `addr`? so that we can support `http://`.

loheagn committed:
storage: add new backend `HttpProxy`

This patch adds a new storage backend `HttpProxy` which can access blobs through a http proxy server.

The http proxy server can be local (using unix socket) or remote (using `https://` or using `http://`).

`HttpProxy` uses two API endpoints to access the blobs:
- `HEAD /path/to/blobs` to get the blob size
- `GET /path/to/blobs` to read the blob

The http proxy server should respect [the `Range` header](https://www.rfc-editor.org/rfc/rfc9110.html#name-range) to compute the offset and length of the blob.

The example config files for this new backend may be:

```jsonc
// for remote usage
{
  "backend": {
    "type": "http-proxy",
    "config": {
      "addr": "http://127.0.0.1:9977",
      "path": "/namespace/<repo>/blobs"
    }
  }
}
```

or

```jsonc
// for local unix socket
{
  "backend": {
    "type": "http-proxy",
    "config": {
      "addr": "/path/to/unix/socket/file"
    }
  }
}
```

There is also a test in `http_proxy.rs` to make sure `HttpProxy` works well, which setups a simple http server and generates a `HttpProxy` backend to get contents from the server.

Signed-off-by: Nan Li <loheagn@icloud.com>

imeoer commented on storage/Cargo.toml:
```diff
@@ -16,6 +16,8 @@ hex = "0.4.3"
 hmac = { version = "0.12.1", optional = true }
 http = { version = "0.2.8", optional = true }
 httpdate = { version = "1.0", optional = true }
+hyper = {version = "0.14.11", optional =  true}
```
Is this necessary? Is `reqwest` haven't supported unix domain socket?

loheagn replied:
Yes. `reqwest` has not supported unix socket yet. The issue https://github.com/seanmonstar/reqwest/issues/39 is still open.

imeoer commented on storage/src/backend/http_proxy.rs:
```diff
@@ -0,0 +1,541 @@
+}
+
+impl BlobReader for HttpProxyReader {
+    // Note! If the blob is compressed, the function will always return 0
```
Is this comment useful?

loheagn replied:
Yes, it's useful. If the blobs are comporessed, we cannnot get the real blob size before we complete the stream reading.

Here is an example in nydus-snapshotter https://github.com/containerd/nydus-snapshotter/blob/3d37ee2f6834ca1626abcd69063f96e1d69d397f/pkg/converter/convert_unix.go#L233.

loheagn commented on api/src/config.rs:
```diff
@@ -427,6 +462,34 @@ pub struct S3Config {
+/// Http proxy configuration information to access blobs.
+#[derive(Clone, Debug, Default, Deserialize, Eq, PartialEq, Serialize)]
+pub struct HttpProxyConfig {
+    /// Address of http proxy server, like `http://xxx.xxx` or `https://xxx.xxx` or `path/to/unix.sock`.
```
Yes. If the user chooses to use unix socket, the `addr` should be a unix path 
string like `/path/to/unix.sock`. The whole backend config may be like:

```json
{
  "backend": {
    "type": "http-proxy",
    "config": {
      "addr": "/path/to/unix/socket/file"
    }
  }
}
```

imeoer replied:
Okay, let's update the comment

loheagn commented on api/src/config.rs:
```diff
@@ -325,6 +349,17 @@ impl BackendConfigV2 {
+
+    /// Get configuration information for http proxy
+    pub fn get_http_proxy_config(&self) -> Result<&HttpProxyConfig> {
+        if &self.backend_type != "http-proxy" {
```
I think both are ok. But for some personal reasons, I prefer `http-proxy`.

loheagn commented on storage/src/backend/http_proxy.rs:
```diff
@@ -0,0 +1,541 @@
+// Copyright 2022 Ant Group. All rights reserved.
```
OK.

imeoer commented on api/src/config.rs:
```diff
@@ -276,6 +278,28 @@ impl BackendConfigV2 {
                 None => return false,
             },
+
+            "http_proxy" => match self.http_proxy.as_ref() {
```
Can we update to `http-proxy`?

loheagn replied:
Oh, my bad...

imeoer commented:
Please also add SOB for the commits.

imeoer commented:
Hi @loheagn, Could also update the doc in [here](https://github.com/dragonflyoss/image-service/blob/master/docs/nydusd.md#use-different-storage-backends)? Thanks!

loheagn commented:
> Hi @loheagn, Could also update the doc in [here](https://github.com/dragonflyoss/image-service/blob/master/docs/nydusd.md#use-different-storage-backends)? Thanks!

No problem. The docs were updated.

imeoer commented on docs/nydusd.md:
````diff
@@ -293,6 +293,51 @@ Currently, the mirror mode is only tested in the registry backend, and in theory
+or
+
+```
+// for remote usage
````
Should be `for local usage` here.

loheagn replied:
fixed :)

changweige commented:
need rebase to master since anolis test script is updated
